### PR TITLE
[Merged by Bors] - feat: missing lemmas about `Finsupp.update`

### PR DESCRIPTION
**Pull request by eric-wieser**
---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Data/Finsupp/Defs.lean [hunk 1]
```diff
@@ -673,6 +678,22 @@ theorem erase_zero (a : α) : erase a (0 : α →₀ M) = 0 := by
   classical rw [← support_eq_empty, support_erase, support_zero, erase_empty]
 #align finsupp.erase_zero Finsupp.erase_zero
 
+theorem update_eq_erase (f : α →₀ M) (a : α) : update f a 0 = f.erase a :=
```
Comment (AntoineChambert-Loir) [inline Mathlib/Data/Finsupp/Defs.lean:681]: wouldn't the name `update_zero_eq_erase` be more explicit?

File: Mathlib/Data/Finsupp/Defs.lean [hunk 2]
```diff
@@ -600,6 +600,11 @@ theorem support_update_ne_zero [DecidableEq α] (h : b ≠ 0) :
   congr; apply Subsingleton.elim
 #align finsupp.support_update_ne_zero Finsupp.support_update_ne_zero
 
+@[simp] theorem update_idem (f : α →₀ M) (a : α) (b c : M) :
```
Comment (AntoineChambert-Loir) [inline Mathlib/Data/Finsupp/Defs.lean:603]: maybe `update_update` would be a more intuitive name.

Comment (eric-wieser) [reply Mathlib/Data/Finsupp/Defs.lean:603]: This matches `Function.update_idem`

File: Mathlib/Data/Finsupp/Defs.lean (hunk 1 above)
Comment (eric-wieser) [reply Mathlib/Data/Finsupp/Defs.lean:681]: Or perhaps `erase_eq_update(_zero)` after flipping the direction?

File: Mathlib/Data/Finsupp/Defs.lean (hunk 2 above)
Comment (AntoineChambert-Loir) [reply Mathlib/Data/Finsupp/Defs.lean:603]: excellent reason to stay by this name!

File: Mathlib/Data/Finsupp/Defs.lean (hunk 1 above)
Comment (AntoineChambert-Loir) [reply Mathlib/Data/Finsupp/Defs.lean:681]: Yes. For a surprising reason, the absence of zero doesn't disturb me in the other direction.

Comment (eric-wieser) [reply Mathlib/Data/Finsupp/Defs.lean]: I've gone for reversing it then

**Comment by AntoineChambert-Loir** [comment]
CI/Build fails at Mathlib/Data/Finsupp/Defs.lean:685:9: 
error:  Finsupp.erase_idem.{u_5, u_1} simp can prove this
I don't how how to fix that except by removing the @[simp] tag before the function.

File: Mathlib/Data/Finsupp/Defs.lean [hunk 3]
```diff
@@ -673,6 +678,22 @@ theorem erase_zero (a : α) : erase a (0 : α →₀ M) = 0 := by
   classical rw [← support_eq_empty, support_erase, support_zero, erase_empty]
 #align finsupp.erase_zero Finsupp.erase_zero
 
+theorem erase_eq_update (f : α →₀ M) (a : α) : f.erase a = update f a 0 :=
+  letI := Classical.decEq α
+  ext fun _ => (Function.update_apply _ _ _ _).symm
+
+@[simp] theorem erase_idem (f : α →₀ M) (a : α) :
```
Comment (eric-wieser) [inline Mathlib/Data/Finsupp/Defs.lean:685]:
```suggestion
-- not `simp` as `erase_of_not_mem_support` can prove this
theorem erase_idem (f : α →₀ M) (a : α) :
```

**Comment by jcommelin** [comment]
Thanks :tada:

bors merge

File: Mathlib/Data/Finsupp/Defs.lean [hunk 4]
```diff
@@ -673,6 +678,23 @@ theorem erase_zero (a : α) : erase a (0 : α →₀ M) = 0 := by
   classical rw [← support_eq_empty, support_erase, support_zero, erase_empty]
 #align finsupp.erase_zero Finsupp.erase_zero
 
+theorem erase_eq_update (f : α →₀ M) (a : α) : f.erase a = update f a 0 :=
```
Comment (YaelDillies) [inline Mathlib/Data/Finsupp/Defs.lean:681]:
```suggestion
theorem erase_eq_update_zero (f : α →₀ M) (a : α) : f.erase a = update f a 0 :=
```
for discoverability. I would even consider
```suggestion
@[simp]
theorem update_zero_eq_erase (f : α →₀ M) (a : α) : f.erase a = update f a 0 :=
```

File: Mathlib/Data/Finsupp/Defs.lean [hunk 5]
```diff
@@ -673,6 +678,23 @@ theorem erase_zero (a : α) : erase a (0 : α →₀ M) = 0 := by
...
+  letI := Classical.decEq α
+  ext fun _ => (Function.update_apply _ _ _ _).symm
+
+-- not `simp` as `erase_of_not_mem_support` can prove this
+theorem erase_idem (f : α →₀ M) (a : α) :
+    erase a (erase a f) = erase a f := by
+  rw [erase_eq_update, erase_eq_update, update_idem]
+
+@[simp] theorem update_erase (f : α →₀ M) (a : α) (b : M) :
+    update (erase a f) a b = update f a b := by
```
Comment (YaelDillies) [inline Mathlib/Data/Finsupp/Defs.lean:691]:
```suggestion
@[simp] theorem update_erase_self (f : α →₀ M) (a : α) (b : M) :
    update (erase a f) a b = update f a b := by
```
and you should have
```
-- @[simp] maybe?
theorem update_erase_of_ne (f : α →₀ M) {a a : α} (ha : a ≠ a') (b : M) :
    update (erase a f) a' b = erase a (update f a' b) := by
```

File: Mathlib/Data/Finsupp/Defs.lean [hunk 6]
```diff
@@ -673,6 +678,23 @@ theorem erase_zero (a : α) : erase a (0 : α →₀ M) = 0 := by
...
+  rw [erase_eq_update, update_idem]
+
+@[simp] theorem erase_update (f : α →₀ M) (a : α) (b : M) :
+    erase a (update f a b) = erase a f := by
```
Comment (YaelDillies) [inline Mathlib/Data/Finsupp/Defs.lean:695]:
```suggestion
@[simp] theorem erase_update_self (f : α →₀ M) (a : α) (b : M) :
    erase a (update f a b) = erase a f := by
```

Comment (eric-wieser) [reply Mathlib/Data/Finsupp/Defs.lean:695]: On `Finset` this is called `erase_insert_eq_erase`

File: Mathlib/Data/Finsupp/Defs.lean (hunk 5 above)
Comment (eric-wieser) [reply Mathlib/Data/Finsupp/Defs.lean:691]: This is called `insert_erase` on `Finset`, so I think the current name is correct.

File: Mathlib/Data/Finsupp/Defs.lean (hunk 4 above)
Comment (eric-wieser) [reply Mathlib/Data/Finsupp/Defs.lean:681]: Did you mean to flip the direction in that last one?

Comment (eric-wieser) [reply Mathlib/Data/Finsupp/Defs.lean:688]: Note that there was already discussion about this in https://github.com/leanprover-community/mathlib4/pull/9316#discussion_r1437765537

File: Mathlib/Data/Finsupp/Defs.lean (hunk 5 above)
Comment (YaelDillies) [reply Mathlib/Data/Finsupp/Defs.lean:698]: `Finset` is a weird comparison. Surely you should rather look at `Function`? I'm looking at `Function.update_same` and `Function.update_noteq` (which are pretty bad names indeed).

Comment (YaelDillies) [reply Mathlib/Data/Finsupp/Defs.lean:698]: Actually, your claim is wrong. `insert_erase` would correspond to `update (erase a f) a (f a) = f`.

Comment (eric-wieser) [reply Mathlib/Data/Finsupp/Defs.lean:698]: Perhaps `update_erase_eq_update` to match the `erase_update_eq_erase` I suggest below?

Comment (YaelDillies) [reply Mathlib/Data/Finsupp/Defs.lean]: That would be good, yes.

File: Mathlib/Data/Finsupp/Defs.lean (hunk 4 above)
Comment (YaelDillies) [reply Mathlib/Data/Finsupp/Defs.lean]: Yep, sorry, I meant it flipped

File: Mathlib/Data/Finsupp/Defs.lean (hunk 5 above)
Comment (eric-wieser) [reply Mathlib/Data/Finsupp/Defs.lean]: I've made that change

**Comment by YaelDillies** [comment]
I think you should
* add `update_erase : update (erase a f) a (f a) = f`
* reverse `erase_eq_update_zero` to `update_zero_eq_erase`

Then I'm happy.

**Comment by YaelDillies** [comment]
I have been convinced otherwise in DMs.

maintainer merge

**Comment by eric-wieser** [comment]
bors merge

**Pull request closed without merging**